cache/ttl/ttl.go

```diff
@@ -0,0 +1,162 @@
+package ttl
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/renproject/kv/db"
+)
+
+var (
+	// PrunePointerTableName is the name of the table where we store the prune pointer.
+	PrunePointerTableName = "ttl_0"
+
+	// PrunePointerKey is the key of the key-value pair which we can use to
+	// query the current prune pointer.
+	PrunePointerKey = "prune_pointer"
+)
+
+type inMemTTL struct {
+	timeToLive    time.Duration
+	pruneInterval time.Duration
+	codec         db.Codec
+	db            db.DB
+}
+
+// New returns a new ttl wrapper over the given database.
+// The underlying database cannot have any database has a prefix of `ttl_`.
+func New(ctx context.Context, database db.DB, timeToLive time.Duration, pruneInterval time.Duration, codec db.Codec) db.DB {
+	ttlDB := inMemTTL{
+		timeToLive:    timeToLive,
+		pruneInterval: pruneInterval,
+		codec:         codec,
+		db:            database,
+	}
+
+	ttlDB.prune(ctx)
+
+	return &ttlDB
+}
+
+func (ttlDB *inMemTTL) Table(name string) db.Table {
+	return ttlDB.db.Table(name)
+}
+
+// Insert implements the `db.table` interface.
+func (ttlDB *inMemTTL) Insert(name string, key string, value interface{}) error {
+
+	// Insert the new data entry into the underlying DB.
+	if key == "" {
+		return db.ErrEmptyKey
+	}
+	if err := ttlDB.db.Insert(name, key, value); err != nil {
+		return err
```

Comment:
Errors should be wrapped with context. For example `return fmt.Errorf("error inserting into ttl: %v", err)` so that when we see errors appear in logs we can track down where they came from.